### PR TITLE
[ECO-2053] Incorporate auditor, testnet campaign feedback

### PR DESCRIPTION
**Pull request by alnoki**
# Source code changes

## API updates

1. Add nonzero `min_output_amount` argument to `swap` for slippage control.
1. Add nonzero `min_lp_coins_out` argument to `provide_liquidity` for rounding
   attack mitigation.
1. Add nonzero `min_quote_out` argument to `remove_liquidity` for rounding
   attack mitigation.
1. Add `balance_as_fraction_of_circulating_supply_before_q64` and
   `balance_as_fraction_of_circulating_supply_after_q64` fields to `Swap` event,
   which require adding type arguments and `ensure_coin_types_initialized` check
   to `simulate_swap_inner` function.

## Revised economic dynamics

1. Enforce winner-take-all donation claim process.
1. Lower state transition threshold to 1,000 APT of cumulative buy pressure.
1. Update market registration function to charge a 4 APT deposit.
1. Update swap to return market registration deposit upon state transition.

## Error case handling

1. Update swap API to calculate FDV/market cap after buy/sell operation to
   prevent unexpected over/underflow errors.
1. Update `cpamm_simple_swap_output_amount` to abort if output amount is 0.
1. Update `simulate_remove_liquidity_inner` to abort if base/quote output amount
   is 0 (no new associated test, due to difficulty of setting up such an edge
   case).
1. Add check for insufficient base balance to `simulate_swap`.

# Docs

1. Update blackpaper source for new economic variables based on lowered
   threshold.
1. Rebuild blackpaper render for docs site.
1. Rebuild Move docs.
1. Update package README with `--override-size-check` command now required to
   publish package.
1. Update docs site values for lowered state transition threshold, with mention
   of new grace period and refundable deposit.
1. Fix typo on docs site that stipulates `cumulative sell pressure` instead of
   `buy pressure`.
1. Relocate blackpaper render location.

# Testing

## Test code updates

1. Lower setup buy amounts for new economic parameters.
1. Update test logic for new winner-take-all donation claim process.
1. Refactor existing test for selling base but not having sufficient balance so
   that it is not triggered by new abort for swap output amount of 0.
1. Add expected failure test for swap where there is no output amount.
1. Add expected failure test for swap min output amount specified as 0.
1. Add expected failure test/minimal happy path for swap min output amount not
   met.
1. Add expected failure test for provide liquidity min lp coins out specified as
   0.
1. Add expected failure test/minimal happy path for provide liquidity min lp
   coins out not met.
1. Add expected failure test for remove liquidity min quote out specified as 0.
1. Add expected failure test/minimal happy path for remove liquidity min quote
   out not met.
1. Update existing tests to support new 4 APT market registration deposit and
   subsequent return.
1. Add expected failure test for registrant unable to afford deposit.
1. Update expected failure swap test case for new base balance check.
1. Update existing tests to support new balance before/after `Swap` fields.
1. Alphabetize test cases for swap simulation.
1. Add test to exercise edge cases for balance as fraction of circulating
   supply before a swap.

## Running tests

1. From in `src/move/emojicoin-dot-fun`:

    ```sh
    aptos move test --dev
    ```

1. Package publishability verified on devnet per Move package README.

Previously the package was tested to 100% coverage, but the coverage tool breaks
for inline functions (https://github.com/aptos-labs/aptos-core/issues/9154) so
re-demonstrating a 100% coverage run requires manually taking out `inline`
keywords as per 6d65d23 and c13dbe2.

As per #181 this PR instead manually adds test cases to hit coverage on new
functionality.

# Checklist

- [x] Did you update relevant documentation?
- [x] Did you add tests to cover new code or a fixed issue?
- [x] Did you check all checkboxes from the linked Linear task?

### CODE REVIEW

File: doc/doc-site/docs/about/how-it-works.md [hunk 1]
```diff
@@ -43,28 +43,35 @@ blockchain.
...
 ### Growing up: exiting the bonding curve
 
 As users buy or sell against the bonding curve, the market capitalization of the
-emojicoin changes. Once 10,000 APT of cumulative sell pressure has pushed the
-bonding curve up to a market capitalization up to 45,000 APT, the emojicoin
-leaves the bonding curve.
+emojicoin changes. Once 1,000 APT of cumulative buy pressure has pushed the
+bonding curve up to a market capitalization up to 4,500 APT, the emojicoin
+leaves the bonding curve and the 4 APT deposit is automatically sent back to the
+market registrant.
```
Comment (xbtmatt) [inline doc/doc-site/docs/about/how-it-works.md:67]:
```suggestion
emojicoin changes. Once 1,000 APT of cumulative buy pressure has pushed the
bonding curve up to a market capitalization of 4,500 APT, the emojicoin
leaves the bonding curve and the 4 APT deposit is automatically sent back to the
market registrant.
```
```suggestion
emojicoin changes. Once 1,000 APT of cumulative buy pressure has pushed the
bonding curve up to a market capitalization up to 4,500 APT, the emojicoin
leaves the bonding curve and the 4 APT deposit is automatically sent back to the
market registrant.
```

Comment (xbtmatt) [reply doc/doc-site/docs/about/how-it-works.md:67]: Or maybe you meant `of up to`? Maybe it's just me but it reads a little awkward 😛

File: src/move/emojicoin_dot_fun/sources/emojicoin_dot_fun.move [hunk 2]
```diff
@@ -543,6 +577,11 @@ module emojicoin_dot_fun::emojicoin_dot_fun {
                 (fdv_end, market_cap_end) =
                     fdv_market_cap(reserves_end, supply_minuend_end);
 
+                // Refund market registration deposit.
+                let RegistrantDeposit { market_registrant, deposit } =
+                    move_from<RegistrantDeposit>(market_address);
+                coin::deposit(market_registrant, deposit);
```
Comment (xbtmatt) [inline src/move/emojicoin_dot_fun/sources/emojicoin_dot_fun.move:583]: There's no way this can get blocked, right? In case of a frozen coin store (I can't remember if you can even freeze APT 😅) or some weird edge case? Only asking because if so it'd lock a market out of ever being able to exit the bonding curve

Maybe `aptos_account::transfer` just in case?

I know that by virtue of having registered a market they must have APT, but just figure might as well make this as flexible as possible

**Review by xbtmatt — APPROVED** [review]
LGTM overall, not sure if the test for min input amount I suggested is necessary or maybe just too difficult

I'll approve at your discretion on when to merge 😄

File: src/move/emojicoin_dot_fun/tests/tests.move [hunk 3]
```diff
@@ -3608,13 +3856,32 @@
...
+        let input_amount = 1_000;
+        mint_aptos_coin_to(USER, input_amount);
+        swap<BlackCatEmojicoin, BlackCatEmojicoinLP>(
+            &get_signer(USER),
+            base_market_metadata().market_address,
+            input_amount,
+            SWAP_BUY,
+            INTEGRATOR,
+            INTEGRATOR_FEE_RATE_BPS,
+            100_000_000_000_000,
```
[the remark this answers is not in the document]
Comment (alnoki) [reply src/move/emojicoin_dot_fun/tests/tests.move]: Addressed in 982c33d

File: doc/doc-site/docs/about/how-it-works.md (hunk 1 above)
Comment (alnoki) [reply doc/doc-site/docs/about/how-it-works.md]: Addressed in 9e7c0d1ad7ea62121f407c8da306b1dbb1baf22c

File: src/move/emojicoin_dot_fun/sources/emojicoin_dot_fun.move (hunk 2 above)
Comment (alnoki) [reply src/move/emojicoin_dot_fun/sources/emojicoin_dot_fun.move:583]: > can get blocked
> ...
> if so it'd lock a market out of ever being able to exit the bonding curve

First of all, I appreciate the intuition on this. Thanks for raising the issue.

> [`aptos_account::transfer`]

That API expects APT at the `source` account, but the refundable deposit is
instead held as a `store`-able field in the new `RegistrantDeposit` so as to
avoid it being accounted for in pool calculations.

Alternatively there is [`aptos_account::deposit_coins`], which performs a check
on [`aptos_account::can_receive_direct_coin_transfers`]. Notably, in the general
case a user could disable the ability to receive direct coin transfers, e.g.
receive coins they have not explicitly registered, via
[`aptos_account::set_allow_direct_coin_transfers`]. So for example there could
be a problem if the refundable deposit was made in `APT` but paid out in
emojicoin, for a registrant who never traded and who turned off direct coin
transfers.

However, in our case it is `APT` in both directions and the registrant
necessarily has to have registered an `AptosCoin`-type `CoinStore` in order to
pay the original market registration deposit (or more specifically, to have had
a proxy for one via FA migration).  Note too that internally,
[`aptos_account::deposit_coins`] calls on `coin::deposit` anyways. So I don't
think that changing APIs would have any functional effect.

> can even freeze APT

This is the real question, and importantly
[`aptos_coin::initialize` destroys the feeze capability].

[`aptos_coin::initialize` destroys the feeze capability]: https://github.com/aptos-labs/aptos-core/blob/f6e43ee347a012ebcbdd7c36933fa99aafeba956/aptos-move/framework/aptos-framework/sources/aptos_coin.move#L54
[`aptos_account::set_allow_direct_coin_transfers`]: https://github.com/aptos-labs/aptos-core/blob/f6e43ee347a012ebcbdd7c36933fa99aafeba956/aptos-move/framework/aptos-framework/sources/aptos_account.move#L145
[`aptos_account::can_receive_direct_coin_transfers`]: https://github.com/aptos-labs/aptos-core/blob/f6e43ee347a012ebcbdd7c36933fa99aafeba956/aptos-move/framework/aptos-framework/sources/aptos_account.move#L182
[`aptos_account::transfer`]: https://github.com/aptos-labs/aptos-core/blob/f6e43ee347a012ebcbdd7c36933fa99aafeba956/aptos-move/framework/aptos-framework/sources/aptos_account.move#L76
[`aptos_account::deposit_coins`]: https://github.com/aptos-labs/aptos-core/blob/f6e43ee347a012ebcbdd7c36933fa99aafeba956/aptos-move/framework/aptos-framework/sources/aptos_account.move#L116

Comment (xbtmatt) [reply src/move/emojicoin_dot_fun/sources/emojicoin_dot_fun.move:583]: Gotcha, awesome thanks for the insight, I thought that they removed the freeze but things have changed and wasn't sure if I'd just missed some crazy change... 🤣

File: src/move/emojicoin_dot_fun/tests/tests.move (hunk 3 above)
[the remark this answers is not in the document]
Comment (xbtmatt) [reply src/move/emojicoin_dot_fun/tests/tests.move]: Awesome 🔥

**Review by xbtmatt — APPROVED** [review]

**Review by CRBl69 — APPROVED** [review]
